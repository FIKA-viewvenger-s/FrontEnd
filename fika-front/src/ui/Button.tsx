import clsx from "clsx";
import type { ButtonHTMLAttributes, FC } from "react";

export interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  variant?: string;
}

const getVariantClassName = (variant: ButtonProps["variant"]) => {
  switch (variant) {
    case "contained":
      return "bg-blue-300";
    case "outlined":
      return "border border-blue-300 bg-white-100";
    default:
      return "";
  }
};

const Button: FC<ButtonProps> = (props) => {
  const { className, disabled, variant, ...rest } = props;
  const variantClassName = getVariantClassName(variant);

  return (
    <button
      className={clsx(
        "w-full rounded-md py-2.5 px-3",
        className,
        variantClassName,
        disabled && "bg-gray-100 text-gray-300"
      )}
      disabled={disabled}
      type="button"
      {...rest}
    />
  );
};

export default Button;
