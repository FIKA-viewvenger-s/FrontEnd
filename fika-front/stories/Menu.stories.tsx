import type { ComponentMeta, ComponentStory } from "@storybook/react";
import Menu from "../src/components/common/menu";

export default {
  component: Menu,
  title: "Menu",
} as ComponentMeta<typeof Menu>;

const Template: ComponentStory<typeof Menu> = (args) => (
  <div className="p-2">
    <Menu {...args} />
  </div>
);

// export const defaulte = Template.bind({});
// defaulte.args = {
//   title: "title이 들어갑니다.",
//   desc: "subtitle이 들어갑니다.",
//   tabItem: ["탭항목", "으로", "사용할", "값이", "들어갑니다."],
// };

// export const noSubTitle = Template.bind({});
// noSubTitle.args = {
//   title: "title이 들어갑니다.",
//   tabItem: ["탭항목", "으로", "사용할", "값이", "들어갑니다."],
// };
