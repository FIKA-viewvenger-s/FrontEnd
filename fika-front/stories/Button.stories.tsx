import type { ComponentMeta, ComponentStory } from "@storybook/react";
import Button from "src/ui/Button";

export default {
  component: Button,
  title: "Button",
} as ComponentMeta<typeof Button>;

const Template: ComponentStory<typeof Button> = (args) => (
  <div className="p-2">
    <Button {...args}>test</Button>
  </div>
);

export const Contained = Template.bind({});
Contained.args = {
  variant: "contained",
};

export const Outlined = Template.bind({});
Outlined.args = {
  variant: "outlined",
};

export const Disabled = Template.bind({});
Disabled.args = {
  disabled: true,
};
