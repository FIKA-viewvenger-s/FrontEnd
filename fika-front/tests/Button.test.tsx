import { render, screen } from "@testing-library/react";
import { composeStories } from "@storybook/react";
import * as stories from "../stories/Button.stories";

describe("index/Button", () => {
  const { Contained, Disabled } = composeStories(stories);

  it("If there is a disable value, it should be disabled.", () => {
    render(<Disabled disabled />);

    const test = screen.getByRole("button");

    expect(test).toBeDisabled();
    expect(test).toMatchSnapshot();
  });

  it("The onClick function should work normally when the button is clicked.", () => {
    const handleClick = jest.fn();

    render(<Contained onClick={handleClick} />);

    const buttonElement = screen.getByRole("button");
    buttonElement.click();

    expect(handleClick).toHaveBeenCalledTimes(1);
    expect(handleClick).toHaveBeenCalled();
    expect(handleClick).toMatchSnapshot();
  });
});
