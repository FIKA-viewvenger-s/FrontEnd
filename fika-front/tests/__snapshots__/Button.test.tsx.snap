// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index/Button If there is a disable value, it should be disabled. 1`] = `
<button
  class="w-full bg-gray-100 text-gray-300"
  disabled=""
  type="button"
>
  test
</button>
`;

exports[`index/Button The onClick function should work normally when the button is clicked. 1`] = `
[MockFunction] {
  "calls": [
    [
      SyntheticBaseEvent {
        "_reactName": "onClick",
        "_targetInst": null,
        "altKey": false,
        "bubbles": true,
        "button": 0,
        "buttons": 0,
        "cancelable": true,
        "clientX": 0,
        "clientY": 0,
        "ctrlKey": false,
        "currentTarget": null,
        "defaultPrevented": false,
        "detail": 0,
        "eventPhase": 3,
        "getModifierState": [Function],
        "isDefaultPrevented": [Function],
        "isPropagationStopped": [Function],
        "isTrusted": false,
        "metaKey": false,
        "movementX": 0,
        "movementY": 0,
        "nativeEvent": MouseEvent {
          "isTrusted": false,
        },
        "pageX": undefined,
        "pageY": undefined,
        "relatedTarget": null,
        "screenX": 0,
        "screenY": 0,
        "shiftKey": false,
        "target": <button
          class="w-full"
          type="button"
        >
          test
        </button>,
        "timeStamp": 1675964204439,
        "type": "click",
        "view": [Window],
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
